src/common/Gateway.sol

```diff
@@ -198,6 +183,31 @@ contract Gateway is Auth, IGateway, Recoverable {
                 failedMessages[centrifugeId][messageHash]++;
                 emit FailMessage(centrifugeId, message, err);
             }
+
+            PoolId poolId = processor.messagePoolId(message);
+            uint256 refunded = (gasleft() - previousGas) * tx.gasprice;
+
+            if (subsidy[poolId] >= refunded) {
+                subsidy[poolId] -= refunded;
+                payable(msg.sender).transfer(refunded);
```

Comment:
We ensure before reaching this line that, the user will always pay more for this execution than the gas refunded, so there is no attack vector here.

Also, because failing messages are enqueued, this action can not be reexecuted with the same parameters.